### PR TITLE
Implement affine methods with tests

### PR DESCRIPTION
**Pull request by sharkAndshark**

### CODE REVIEW

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
+            {
+                const string ewkt = "LINESTRING (50 160, 50 50, 100 50)";
+                db.TestGeometries.Value(g => g.Id, 1).Value(p => p.Geometry, () => GeometryInput.STGeomFromEWKT(ewkt)).Insert();
+                var result = db.TestGeometries
+                    .Where(g => g.Id == 1)
+                    .Select(g => g.Geometry.STRotate(Math.PI))
+                    .Select(g => GeometryOutput.STAsEWKT(g))
+                    .Single();
+                Assert.AreEqual(result, "LINESTRING(-50.00000000000002 -160,-50.00000000000001 -49.99999999999999,-100 -49.999999999999986)");
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: Test fails

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
+            {
+                const string ewkt = "LINESTRING(1 1, 2 2)";
+                db.TestGeometries.Value(g => g.Id, 1).Value(p => p.Geometry, () => GeometryInput.STGeomFromEWKT(ewkt)).Insert();
+                var result = db.TestGeometries
+                    .Where(g => g.Id == 1)
+                    .Select(g => g.Geometry.STScale(GeometryInput.STGeomFromEWKT("POINT(2 2)"), GeometryInput.STGeomFromEWKT("POINT(1 1)")))
+                    .Select(g => GeometryOutput.STAsEWKT(g))
+                    .Single();
+                Assert.AreEqual(result, "LINESTRING(1 1,3 3)");
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: The Assert method signature is
`AreEqual(expected, actual)`
so it should look as
` Assert.AreEqual("LINESTRING(1 1,3 3)", result);`

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
+            {
+                const string ewkt = "LINESTRING (50 160, 50 50, 100 50)";
+                db.TestGeometries.Value(g => g.Id, 1).Value(p => p.Geometry, () => GeometryInput.STGeomFromEWKT(ewkt)).Insert();
+                var result = db.TestGeometries
+                    .Where(g => g.Id == 1)
+                    .Select(g => g.Geometry.STRotate(Math.PI / 6.0, 50, 160))
+                    .Select(g => GeometryOutput.STAsEWKT(g))
+                    .Single();
+                Assert.AreEqual(result, "LINESTRING(50 160,104.99999999999999 64.73720558371174,148.30127018922192 89.73720558371173)");
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: Test fails
Actual `result` is 
"LINESTRING(50 160,105 64.7372055837117,148.301270189222 89.7372055837117)"
Can the rounding depends on PostgreSQL/PostGIS version? 
I use PostgreSQL 9.6.1, PostGIS 3.0.2

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
+            {
+                const string ewkt = "LINESTRING (50 160, 50 50, 100 50)";
+                db.TestGeometries.Value(g => g.Id, 1).Value(p => p.Geometry, () => GeometryInput.STGeomFromEWKT(ewkt)).Insert();
+                var result = db.TestGeometries
+                    .Where(g => g.Id == 1)
+                    .Select(g => g.Geometry.STRotate(-1.0 * Math.PI / 3.0, g.Geometry.STCentroid()))
+                    .Select(g => GeometryOutput.STAsEWKT(g))
+                    .Single();
+                Assert.AreEqual(result, "LINESTRING(116.42245883568916 130.67207346706593,21.15966441940092 75.67207346706593,46.15966441940093 32.370803277844)");
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: Test fails, due to rounding errors after converting to EWKT.
How about returning `Geometry` and using `Assert.AreEqual` with precision tolerance ?

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs (hunk 3 above)
Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: maybe.
i use postgis 3.1.0... potgresPostgreSQL 13.1 ; 
i will get my pg/postgis version sync to 9.6.1&3.0.2 and try fix the precision of geometry output

Comment (apdevelop) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: Would be better to use `Assert.AreEqual` with precision tolerance on returned `Geometry` (not WKT/EWKT), instead of locking on specific versions.

File: LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs (hunk 4 above)
Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: how about geom.EqualsNormalized?
```
var expectedGeom = new NetTopologySuite.IO.WKTReader().Read("LINESTRING(-1 -2 3,-1 -4 3)");
var result = db.DoSomeThing().Single();
Assert.True(expectedGeom.EqualsNormalized(result);
```

Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: get what u mean .how about
```
var expected = new NetTopologySuite.IO.WKTReader().Read("some geom text.....");
var result = db.DoSomeThing().Single();
var isEqual = expectedGeom.Normalized().EqualsExact(result.Normalized(),0.001);// suggestion of tolerance value?
Assert.True(isEqual);
```

Comment (apdevelop) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: Something like that:
```
var result = db.TestGeometries
                     .Where(g => g.Id == 1)
                      .Select(g => g.Geometry.STRotate(Math.PI))
                      .Single() as NTSG.LineString;
Assert.AreEqual(-50.0, result.GetPointN(1).X, 1.0E-9); // Actually -50.000000000000007
```
The purpose of these tests is checking proper declaration of `Sql.Function`, not checking PostGIS spatial methods  implementation and accuracy.

Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: : D  
is it ok that some affine tests like this:
```
class AffineTransformationsTests : TestsBase
{
//some tests here..

        private void CheckLineEquals(double[][] expected,NTSG.LineString result,double tolerance)
        {
            var pointsCount = expected.Length;
            var dimensionCount = expected[0].Length;

            for (int i = 0; i < pointsCount; i++)
            {
                for (int j = 0; j < dimensionCount; j++)
                {
                    var coordnates = result.GetCoordinateN(i);
                    var exp = expected[i][j];
                    var res = result.GetCoordinateN(i).CoordinateValue[j];
                    Assert.AreEqual(exp,res,tolerance);
                }
            }
        }
} 
```
        [Test]
        public void TestSTAffine3D()
        {
            var expected = new double[][]
                {
                    new[]{-1.0,-2.0,3.0},
                    new[]{-1.0,-4.0,3.0}
                };
            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
            {
                const string ewkt = "LINESTRING(1 2 3, 1 4 3)";
                db.TestGeometries.Value(g => g.Id, 1).Value(p => p.Geometry, () => GeometryInput.STGeomFromEWKT(ewkt)).Insert();
                var result = db.TestGeometries
                    .Where(g => g.Id == 1)
                    .Select(g =>
                        g.Geometry.STAffine(Math.Cos(Math.PI),
                            -1 * Math.Sin(Math.PI), 0, Math.Sin(Math.PI), Math.Cos(Math.PI), 0, 0, 0, 1, 0, 0, 0)
                    )
                    .Single() as NTSG.LineString;
                ;
                CheckLineEquals(expected,result,1.0E-8);
            }
        }
```
```

Comment (apdevelop) [reply LinqToDBPostGisNetTopologySuite.Tests/AffineTransformationsTests.cs]: I think it's OK for such complex spatial methods. But always remember to keep test methods as simple as possible. It's more important to check corner cases, like checking `null` `double?` result from `STArea()` for `null` input.

**Comment by apdevelop** [comment]
Thank you for the completed PR!
